core/hashbuffer.go

```diff
@@ -0,0 +1,138 @@
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [hashBufferElems]common.Hash
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The block number for head (the most recent block)
+	tailNumber uint64 // The block number for tail (the oldest block)
+}
+
+// newHashStorage creates a new storage with a header in it.
+// Since we take a header here, a hash storage can never be empty.
+// This makes things easier later on (in Set)
+func newHashStorage(header *types.Header) *hashBuffer {
+	return &hashBuffer{
+		headNumber: header.Number.Uint64(),
+		tailNumber: header.Number.Uint64(),
+		data:       [hashBufferElems]common.Hash{header.Hash()},
+	}
+}
+
+// Get locates the hash for the requested number
+func (hs *hashBuffer) Get(number uint64) (common.Hash, bool) {
+	if !hs.has(number) {
+		return common.Hash{}, false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.data[index], true
+}
+
+// has returns if the storage has a hash for the given number
+func (hs *hashBuffer) has(number uint64) bool {
+	return number <= hs.headNumber && number >= hs.tailNumber
+}
+
+// Contains checks if the hash at the given number matches the expected
+func (hs *hashBuffer) Contains(number uint64, expected common.Hash) bool {
+	if !hs.has(number) {
+		return false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.data[index] == expected
+}
+
+// Newest returns the most recent (number, hash) stored
+func (hs *hashBuffer) Newest() (uint64, common.Hash) {
+	return hs.headNumber, hs.data[hs.head]
+}
+
+// Oldest returns the oldest (number, hash) found
+func (hs *hashBuffer) Oldest() (uint64, common.Hash) {
+	distance := hs.headNumber - hs.tailNumber
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.tailNumber, hs.data[index]
+}
+
+// Set inserts a new header (hash) to the storage.
+// If
+// a) Header already exists, this is a no-op
+// b) Number is occupied by other header, the new header replaces it, and also
+// truncates any descendants
+//
+// If the new header does not have any ancestors, it replaces the entire storage.
+func (hs *hashBuffer) Set(header *types.Header) {
+	var (
+		number = header.Number.Uint64()
+		index  uint64
+		hash   = header.Hash()
+	)
+	if hs.Contains(number-1, header.ParentHash) {
+		if hs.headNumber >= number {
+			distance := hs.headNumber - number
+			index = (hs.head + hashBufferElems - distance) % hashBufferElems
+			if bytes.Compare(hs.data[index][:], hash[:]) == 0 {
+				return
+			}
+			// Continue by replacing this number and wipe descendants
+		} else {
+			// head is parent of this new header - regular append
+			index = (hs.head + 1) % hashBufferElems
+		}
+	} else {
+		// This should not normally happen, and indicates a programming error
```

Comment:
Doesn't a side-chain import also trigger this path? If so, it's not a programming error.

Reply:
Only if the sidechain import is old, older than 288 blocks. This 'not _normally_ happen'. We only add hashes of canon blocks, from blockchain.go, either in `insertChain` or `reorg`

Reply:
There is another case of this:

* We initialize the `hashStorage` with genesis header
* We get a header from the network but it's not consecutive with genesis.

Maybe we should consider initialize the `hashStorage` with `latestHeader`?

Reply:
>     We get a header from the network but it's not consecutive with genesis.

Why would we ever import a header that is not consecutive with genesis? And if we were to do so, why is it not correct to wipe ancestors?